xdsl/builder.py

```diff
@@ -99,14 +107,160 @@ def region(input: Callable[[Builder], None]) -> Region:
+
+        For regions that have inputs or outputs:
+        ```
+        @Builder.implicit_region(input_types)
+        def func(args: tuple[BlockArgument, ...]) -> None:
+            ...
+        ```
+
+        For regions that don't have inputs or outputs:
+        ``` python
+        @Builder.implicit_region
+        def func() -> None:
+            ...
+        ```
+        """
+        ...
+
+    @overload
+    @staticmethod
+    def implicit_region(input: Callable[[], None]) -> Region:
+        ...
+
+    @staticmethod
+    def implicit_region(
+        input: Sequence[Attribute] | ArrayAttr[Attribute]
+        | Callable[[], None]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region] | Region:
+        if isinstance(input, Callable):
+            return Builder._implicit_region_no_args(input)
+        else:
+            return Builder._implicit_region_args(input)
+
+
+# Implicit builders
+
+
+@dataclass
+class _ImplicitBuilderStack(threading.local):
+    """
+    Stores the stack of implicit builders for use in @Builder.implicit_region, empty by
+    default. There is a stack per thread, guaranteed by inheriting from `threading.local`.
+    """
+    stack: list[Builder] = field(default_factory=list)
+
+    def push(self, builder: Builder) -> None:
+        self.stack.append(builder)
+
+    def get(self) -> Builder | None:
+        if len(self.stack):
+            return self.stack[-1]
+
+    def pop(self, builder: Builder) -> Builder:
+        popped = self.stack.pop()
+        assert popped is builder
+        return popped
+
+
+@dataclass
+class _ImplicitBuilder(contextlib.AbstractContextManager[None]):
+    """
+    Stores the current implicit builder context, consisting of the stack of builders in
+    the current thread, and the current builder.
+    """
+
+    _stack: ClassVar[_ImplicitBuilderStack] = _ImplicitBuilderStack()
+
+    builder: Builder
+
+    def __enter__(self) -> None:
+        type(self)._stack.push(self.builder)
+
+    def __exit__(self, __exc_type: type[BaseException] | None,
+                 __exc_value: BaseException | None,
+                 __traceback: TracebackType | None) -> bool | None:
+        type(self)._stack.pop(self.builder)
+
+    @classmethod
+    def get(cls) -> Builder | None:
+        return cls._stack.get()
+
 
 _CallableRegionFuncType: TypeAlias = Callable[
     [Builder, tuple[BlockArgument, ...]], None]
+_CallableImplicitRegionFuncType: TypeAlias = Callable[
+    [tuple[BlockArgument, ...]], None]
+
+
+def _op_init_callback(op: Operation):
+    if (b := _ImplicitBuilder.get()) is not None:
+        b.insert(op)
+
+
+def _override_operation_post_init() -> None:
+    old_post_init = Operation.__post_init__
+
+    def new_post_init(self: Operation) -> None:
+        old_post_init(self)
+        _op_init_callback(self)
+
+    Operation.__post_init__ = new_post_init
+
+
+# set up the operation callback for implicit construction
+_override_operation_post_init()
```

Comment:
Yes, I prefer this a lot!
I am not quite sure what happens in a future where this mechanism becomes more complex and this might be imported, and therefore run, multiple times. But for the time being, I am happy with this.

Reply:
Yeah, let's see how this goes. Maybe we'll find a neater way to do it with a macro system rather than the hacky callback injection, but it feels like the API is right already, and it should be an implementation-only thing, that the client won't need to worry about.